### PR TITLE
add lexer option to tokenize whitespaces under version DMDLIB

### PR DESCRIPTION
**Pull request by sorin-gabriel**
The changes are motivated by the need to improve the versatility of the lexer and other parts of the compiler such that dmd will eventually replace 3rd party libraries like `libdparse`.

This pull request in particular adds the option to make tokens out of whitespaces and implements the range interface over the lexer. Behaviour is unchanged if version `DMDLIB` is not used. Otherwise, the lexer should be viewed as a range and the range-specific methods should be used; the `popFront()` call inside the constructor is needed to skip over the first token (which is always initialized to `TOK.reserved` and it would have been skipped by calling `nextToken()`).

Will add tests soon.

### CODE REVIEW

**Comment by WalterBright** [comment]
Speed is crucial to the lexer. Any changes to this should be speed tested.

Also, the lexer changes very little over time. It is not a burden to make a lexer that handles whitespace differently.

**Comment by RazvanN7** [comment]
Currently, all the additions are under `version(DMDLIB)`, so this code will not be included in the release version of the compiler.

File: src/dmd/lexer.d [hunk 1]
```diff
@@ -141,6 +147,43 @@ class Lexer
...
+
+        TOK front() const @property
+        {
+            return token.value;
+        }
+
+        void popFront()
+        {
+            nextToken();
+        }
```
Comment (MoonlightSentinel) [inline src/dmd/lexer.d:173]: The addition of a range interface should be done in another commit / PR

File: src/dmd/lexer.d [hunk 2]
```diff
@@ -141,6 +147,43 @@ class Lexer
...
+        {
+            TOK[] tokens;
+            while (!empty())
+            {
+                tokens ~= front();
+                popFront();
+            }
+            return tokens;
+        }
+    }
```
Comment (MoonlightSentinel) [inline src/dmd/lexer.d:185]: Remove, this duplicates `std.array.array`

File: src/dmd/lexer.d [hunk 3]
```diff
@@ -141,6 +147,43 @@ class Lexer
...
     }
 
+    version (DMDLIB)
+    {
+        this(const(char)* filename, const(char)* base, size_t begoffset, size_t endoffset,
+            bool doDocComment, bool commentToken, bool whitespaceToken)
+        {
+            this(filename, base, begoffset, endoffset, doDocComment, commentToken);
+            this.whitespaceToken = true;
+            this.popFront();
```
Comment (MoonlightSentinel) [inline src/dmd/lexer.d:157]: This introduces an inconsistency between both ctors - instances created by the existing ctor yield `Token.init` as their first element

File: src/dmd/lexer.d [hunk 4]
```diff
@@ -590,8 +665,12 @@ class Lexer
...
                     if (commentToken)
                     {
-                        p++;
-                        endOfLine();
+                        version (DMDLIB) {}
+                        else
+                        {
+                            p++;
+                            endOfLine();
+                        }
```
Comment (MoonlightSentinel) [inline src/dmd/lexer.d:673]: This should only be skipped when `whitespaceToken` is `true`?

**Review by MoonlightSentinel — DISMISSED** [review]
Please add a test case, see e.g. [this file](https://github.com/dlang/dmd/blob/master/test/unit/lexer/location_offset.d).

**Comment by ljmf00** [comment]
> Speed is crucial to the lexer. Any changes to this should be speed tested.
> 
> Also, the lexer changes very little over time. It is not a burden to make a lexer that handles whitespace differently.

I would suggest adding the option to the lexer template instead of making it specific to DMDLIB. This way we don't hurt the performance of someone who uses the DMD lexer for performance reasons.

File: src/dmd/tokens.d [hunk 5]
```diff
@@ -56,6 +56,7 @@ enum TOK : ubyte
     pragma_,
     typeid_,
     comment,
+    whitespace,
```
Comment (ibuclaw) [inline src/dmd/tokens.d:59]: I'd suggest putting new members at the end of the enum.

File: src/dmd/lexer.d (hunk 4 above)
Comment (sorin-gabriel) [reply src/dmd/lexer.d:673]: Yes, but actually no. Suppose we have
```
// some comment(implicit '\n')
some statement
```
and the current iteration is k.

Without `version DMDLIB`:
k / TOK.comment and p is incremented past the '\n'
k+1 / TOK.statement

With `version DMDLIB`, _as it is now_:
- if `whitespaceToken` is `true`:
k / TOK.comment and p is pointing to the '\n'
k+1 / TOK.whitespace
k+2 / TOK.statement
- if `whitespaceToken` is `false`:
k / TOK.comment and p is pointing to the '\n'
k+1 / hits `case '\n'`, (but we don't lex whitespaces) increments p and continues
k+2 / TOK.statement

If we had something like
```
version (DMDLIB)
{
    if (whitespaceToken == false)
    {
        p++;
        endOfLine();
    }
}
else
{
    p++;
    endOfLine();
}
```
- if `whitespaceToken` is `true`:
same as above
- if `whitespaceToken` is `false`:
k / TOK.comment and p is incremented past the '\n'
k+1 / TOK.statement

Basically, the difference boils down to what we would prefer: one more if-statement to check or one more case to match.

File: src/dmd/lexer.d (hunk 2 above)
Comment (sorin-gabriel) [reply src/dmd/lexer.d:185]: Removed

File: src/dmd/lexer.d (hunk 3 above)
Comment (sorin-gabriel) [reply src/dmd/lexer.d:157]: Removed call to `popFront()`.

File: src/dmd/tokens.d (hunk 5 above)
Comment (sorin-gabriel) [reply src/dmd/tokens.d:59]: To me it makes more sense to have `whitespace` near `comment`. I would keep it like this, unless we see problems popping up because of it.

Comment (ibuclaw) [reply src/dmd/tokens.d:59]: It's:
1. The huge diff it generates, and
2. The invalidating of all comments in the hand maintained headers (yeah, yeah, those `// 54` comments should be removed from the headers).

**Review by ibuclaw — DISMISSED** [review]
> Speed is crucial to the lexer. Any changes to this should be speed tested.
> 
> Also, the lexer changes very little over time. It is not a burden to make a lexer that handles whitespace differently.

There's also something that may resemble an ImportC bug fix in this PR somewhere, so some parts shouldn't be hiding behind `version (DMDLIB)`.

@WalterBright e.g:  This wrong-code is accepted by dmd:
```
#
    pragma
pack
    (
2
    )
```
See https://issues.dlang.org/show_bug.cgi?id=22812 and https://issues.dlang.org/show_bug.cgi?id=22814

**Comment by RazvanN7** [comment]
@ibuclaw I think that fixing an importC bug is something that is beyond the scope of this PR. From my perspective, this addition is a clear improvement for dmd-as-a-lib since it aligns the interface to what existing tools expect. Moreover it does not impact the performance of the compiler in any way. If parts of it can be reused to fix existing bugs that's fine, but that shouldn't delay this addition.

@sorin-gabriel However, the issue of adding the token at the end of the enum still remains. So please fix that.

@ibuclaw are there any specific changes that you request for this PR?

File: src/dmd/lexer.d (hunk 1 above)
Comment (sorin-gabriel) [reply src/dmd/lexer.d]: Will do

File: src/dmd/tokens.d (hunk 5 above)
Comment (sorin-gabriel) [reply src/dmd/tokens.d]: Adding `whitespace` to the very end creates problems with the `__traits(getMember..)` / `__traits(allMembers..)`:
[https://auto-tester.puremagic.com/show-run.ghtml?projectid=1&runid=5130725&isPull=true](https://auto-tester.puremagic.com/show-run.ghtml?projectid=1&runid=5130725&isPull=true)
Looks like a different bug (maybe the fields beginning with `__` are not counted correctly).

**Comment by RazvanN7** [comment]
@MoonlightSentinel @ibuclaw do you have any other objections to this PR?

File: src/dmd/tokens.h [hunk 6]
```diff
@@ -283,6 +284,7 @@ enum class TOK : unsigned char
     stdcall,
     attribute__,
 
+
```
Comment (RazvanN7) [inline src/dmd/tokens.h:287]: unrelated change, undo please

File: src/dmd/tokens.d [hunk 7]
```diff
@@ -880,6 +883,7 @@ extern (C++) struct Token
         TOK.__declspec     : "__declspec",
         TOK.__stdcall      : "__stdcall",
         TOK.__attribute__  : "__attribute__",
+
```
Comment (RazvanN7) [inline src/dmd/tokens.d:886]: undo

File: src/dmd/tokens.d [hunk 8]
```diff
@@ -273,6 +274,7 @@ enum TOK : ubyte
     __declspec,
     __stdcall,
     __attribute__,
+
```
Comment (RazvanN7) [inline src/dmd/tokens.d:277]: undo

**Review by RazvanN7 — APPROVED** [review]
